tests/main/snapd-snap/task.yaml

```diff
@@ -1,27 +1,143 @@
+        tests.cleanup defer snap remove lxd
+
+        echo "Setup the lxd snap"
+        snap set lxd waitready.timeout=240
+        lxd waitready
+        lxd init --auto
+
+        echo "Setting up proxy for lxc"
+        if [ -n "${http_proxy:-}" ]; then
+            lxd.lxc config set core.proxy_http "$http_proxy"
+        fi
+        if [ -n "${https_proxy:-}" ]; then
+            lxd.lxc config set core.proxy_https "$http_proxy"
+        fi
+
+        echo "Switch the core snap back due to spread system prepare shenanigans"
+        # When we run spread, we will repack the core snap that is installed on
+        # the system to match the binaries that we built on this system so that
+        # we keep executing the stuff from the branch, even if re-exec is 
+        # enabled unconditionally for all spread tests. This presents a problem
+        # for snapcraft, which downloads the current snap file from snapd 
+        # directly using the /v2/snaps/{name}/file endpoint, which will just 
+        # send over the modified file without indicating at all that the current
+        # file is modified from the installed revision (indeed snapd reports to
+        # snapcraft that it is running the specific edge revision from the 
+        # store, and not x1 as it would in devmode for example which snapcraft
+        # correctly handles).
+        # As such, we need to switch the modified core snap on the system back 
+        # with the original core snap on the system that is unmodified and will
+        # match the expected assertions that snapcraft calculates
+
+        # the easy way is to trick snapd is to just mount back the original file
+        # on top of the repacked one, then snapd will just serve the correct
+        # file over HTTP to snapcraft when it asks for it
+
+        CORE_SNAP_REV=$(snap list core | grep -v Name | awk '{print $3}')
+        CORE_SNAP="/var/lib/snapd/snaps/core_${CORE_SNAP_REV}.snap"
+        ORIG_CORE_SNAP=${CORE_SNAP}.orig
+        
+        mount --bind "$ORIG_CORE_SNAP" "$CORE_SNAP"
+        tests.cleanup defer umount "$CORE_SNAP"
+    fi
 
 restore: |
+    # shellcheck disable=SC2164
     cd "$PROJECT_PATH"
-    echo "Cleanup build artifacts"
-    not command -v snapcraft >/dev/null || snapcraft clean --destructive-mode
-    echo "Cleanup the build snapd snap"
-    rm -f snapd_spread-test.snap
+    tests.cleanup restore
 
 execute: |
+    # only run the destructive mode variant on xenial
+    if ! os.query is-xenial && [ "$SNAPCRAFT_BUILD_ENVIRONMENT" = "host" ]; then
+        echo "Skipping destructive mode build on non-xenial"
+        exit 0
+    fi
+
     # shellcheck disable=SC2164
     cd "$PROJECT_PATH"
-    echo "Build the snap natively (with --destructive-mode)"
-    snapcraft snap --destructive-mode --output=snapd_spread-test.snap
+    echo "Build the snap"
+    snap run snapcraft snap --output=snapd_spread-test.snap
+    tests.cleanup defer rm -f snapd_spread-test.snap
```

Comment:
this is not needed, the PWD is automatically restored

Reply:
no, actually this is needed, since we are not in PWD, we are in $PROJECT_PATH which is not restored afaict

Reply:
ah, in that case, should we use the full path for the snap?

Reply:
I refactored this see my other comment